### DISCUSSION TITLE
Heterogeneous illumination correction - Is there any specific function already available?

### DISCUSSION THREAD

**Discussion by lscaldas** [hyperspy/hyperspy#3394] [Q&A] [answered]
I have been using Hyperspy with good success to work with a spectromicroscopy dataset (PEEM/NEXAFS). However, one of my major problems when dealing with PCA is the presence of heterogeneous illumination that can't be corrected instrumentally. Simply put, the incident X-ray beam moves across the different images/ photon energies. So I cannot simply create a blurry version of this image and use as flat field, because the center of the beam moves. I tried modelling the beam movement, and applying this correction, but the result was not satisfactory. So far I have not been able to isolate this influence in an individual component (I tried PCA and BSS) to be removed. I can get a component, normally the first, that doesnt contain the absorption peak of the element of interest, but still the other components have some leftover effect of the beam, so I dont consider this a perfect isolation of the component. 

Here is an example of how the beam is shaped for a given energy:
![31_0615h_fullimage_0](https://github.com/hyperspy/hyperspy/assets/42618468/e31dbfa9-9818-4d26-9307-8f4f4a1289b3)

And here is why I believe it is possible. The nanoparticles in display here, are the only ones with the element being probed. The substrate itself has only the background intensity. As you can see the bottom left background is brighter than the top right background:
![0615h_plus_i_0030](https://github.com/hyperspy/hyperspy/assets/42618468/eeca3ba8-9b18-409f-883e-0b8d5af81565)

Maybe there is a functionality already implemented here, or perhaps someone has a suggestion on how to solve this issue, through hyperspy or another way. If someone could help, I would be very grateful.

**Comment by ericpre** [comment, marked as answer]
There isn't any specific function available for that.
Generally, if you can correct each image individually, then you can use the [map](https://hyperspy.org/hyperspy-doc/current/reference/api.signals/BaseSignal.html#hyperspy.api.signals.BaseSignal.map) method to apply the function to each navigation position.

Idea of potential correction:
- use function that would remove the low frequency component of the images and apply this function using [map](https://hyperspy.org/hyperspy-doc/current/reference/api.signals/BaseSignal.html#hyperspy.api.signals.BaseSignal.map)
  - as you said, use a blurry version of each image and subtract that.
  - Try https://pyxem.readthedocs.io/en/stable/examples/processing/background_subtraction.html#sphx-glr-examples-processing-background-subtraction-py. These are more for images containing peaks but it may still work for the features that you have in your data.
  - If you have an image of the beam shape to be use as a template, use cross correlation to find the position and subtract that but it may need some thinking to find a good way to normalise the template.
  - subtract background using the [rolling ball function from scikit-image](https://scikit-image.org/docs/stable/auto_examples/segmentation/plot_rolling_ball.html#sphx-glr-auto-examples-segmentation-plot-rolling-ball-py).
  - Use one of the 2D algorithms implemented in [pybaselines ](https://pybaselines.readthedocs.io/en/latest/algorithms_2d/index.html). These may be very slow...
  - any function that does a good job at subtracting a low varying background.
- The [rolling ball function from scikit-image work with 3 dimensions](https://scikit-image.org/docs/stable/auto_examples/segmentation/plot_rolling_ball.html#higher-dimensions) and it may be worth comparing with its use when apply to each image individually. It would work as follow:
  - [unfold](https://hyperspy.org/hyperspy-doc/current/reference/api.signals/BaseSignal.html#hyperspy.api.signals.BaseSignal.unfold) the signals
  - apply directly the function to the signal
  - [fold](https://hyperspy.org/hyperspy-doc/current/reference/api.signals/BaseSignal.html#hyperspy.api.signals.BaseSignal.fold) the signal back
- Use Fourier filtering to filter the very low frequency components, it may not work very well because of the image sampling not being suitable for that but I mentioned it here in the interest of completeness! This would consist in taking the [FFT](https://hyperspy.org/hyperspy-doc/current/reference/api.signals/BaseSignal.html#hyperspy.api.signals.BaseSignal.fft), setting the low frequency component to zero and taking the inverse FFT

To understand what kind of correction would make sense, could you explain in more details what the source of the beam movement, and why is it not possible to correct experimental.

**Reply by CSSFrancis** [reply]
>   - If you have an image of the beam shape to be use as a template, use cross correlation to find the position and subtract that but it may need some thinking to find a good way to normalise the template.

The match_template function in skimage that we use in pyxem is exactly designed for variable illumination. 

https://pyxem.readthedocs.io/en/stable/examples/processing/template_matching.html#sphx-glr-examples-processing-template-matching-py

**Comment by lscaldas** [comment]
First, I would like to thank both off you to offer some nice insight and options to tackle this issue. I end up doing the following, and I think it worked, although I need to test with more data:

1) Gaussian blur on each image (100).
The issue of applying the gaussian blur image directly to correct the data as a flat field, is that by doing this I am not only removing the beam inhomogeneity, but also the average spectra of the stack of images.  So to counterattack this, you need to separate what is the i0, so the effect of a light source center (in this case a synchrotron low energy X-ray) moving away from the image, from the signal generated from the X-ray absorption of the object of interest, in my case nanoparticles. To do this I normalized each image by dividing by the maximum intensity of each image. But to avoid having some artifacts, I:
2) Used curve_fit from scipy to fit a gaussian simulating my beam for each image.
3) Took the amplitude of each fit, and used to normalize the gaussian blur flat field images of step 1. 
4) I then applied the normalized flat field to correct my main data. This removed the beam inhomogeneity across the lateral space (x, y), and across the energy space (so no decay caused by a beam moving away), while maintaining the peak features expected for my chemical species.
5) I then applied decomposition and bss, and it was possible isolate components that look like chemical species without the influence of the beam movement. 
6) When I compared doing this process with just using the data without the correction, the number of components went from 5 (uncorrected) to 3 (illumination corrected), and the uncorrected data was, as I said before, not able to isolate the beam effect on a single component, white the illuminated corrected didnt show components that had intensity decays.

I dont know yet if this is a general solution, or if I didn't create artifacts during this process. I will know in the following weeks. But thank you very much for the suggestions, they helped, and I marked as answered.

Regarding why the beam moves, is just because of a mechanical movement of a mirror when the photon energy varies in this specific beamline. It is not a massive movement, but something like 1 µm per 50 eV, but since my NPs are 20 nm, and the NEXAFS go over this energy range,  and the beam is also not a perfect spot, it counts.  In thesis it is possible to program some counter measure, but it would demand some time/personal/cost investment, which is not priority now. And the data in question can not be retaken. 

Now on a tangent note, one of my issues is how to perform BSS with non negative components. From what I understood, I can do decomposition with NMF, but I don't know if I can do with BSS as well. Do you know? I am having more luck finding components that resemble chemical species with the BSS than with PCA alone, considering it (NEXAFS) is similar to EELS, I imagine this is expected. The issue is that sometimes I have negative components (sometimes that look like the fingerprint, but flipped). Also do you know if it is possible to feed the BSS calculation with a known component, like for instance, lets say that I know one chemical species is there, and I have the fingerprint of it. But i don't have a fingerprint of the second chemical species, and I would like the BSS to arrive to this. Is this doable in hyperspy?

**Reply by ericpre** [reply]
You can reverse [bss components](https://hyperspy.org/hyperspy-doc/current/reference/api.signals/BaseSignal.html#hyperspy.api.signals.BaseSignal.reverse_bss_component).
For your other questions, you could try to use MCR, which can be used in hyperspy by doing https://github.com/hyperspy/hyperspy/pull/2172#issuecomment-688781633.
